File: Sean的leetcode/Leetcode/Missing Ranges.cpp
class Solution {
public:
    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        vector<string> res;
        int pre = lower - 1;
        int n = nums.size();

        //note: i is from 0 to n
        for (int i = 0; i <= n; i++)
        {
            int curr = i < n ? nums[i] : upper + 1;

            if (curr == pre + 2)
                res.push_back(to_string(pre+1));
            else if (curr > pre + 2)
                res.push_back(to_string(pre+1)+"->"+to_string(curr-1));
            pre = curr;
        }
        return res;
    }
};

File: MyLeetCode/python/Missing Ranges.py
###
# Array
# Time Complexity: O(n)
# Space Complexity: O(n)
###
class Solution(object):
    def findMissingRanges(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: List[str]
        """
        pre = lower - 1
        nums.append(upper+1)
        re = []
        for n in nums:
            if n == pre + 2:
                re.append(str(pre+1))
            elif n > pre + 2:
                re.append(str(pre+1) + "->" + str(n-1))
            pre = n     
        return re
        
        